### PR TITLE
Fix build in macOS XCode 12

### PR DESCRIPTION
**Pull request by JasonInOttawa**
Add compile flags to allow implicit function declarations in libk2pdfopt and minizip

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader-base/1198)

### CODE REVIEW

File: thirdparty/minizip/CMakeLists.txt [hunk 1]
```diff
@@ -13,14 +13,23 @@ assert_var_defined(AR)
...
 # FIXME: Even then, something in said branch seems to upset MuPDF with our custom patch as-is, so keep using the old code...
 #set(PATCH_CMD "${KO_PATCH} ${CMAKE_CURRENT_SOURCE_DIR}/minizip-libaes-makefile.patch")
 
 set(PATCH_CMD sh -c "${ISED} \"s|^CC=|#CC=|g\" aes/Makefile && ${ISED} \"s|^CFLAGS=|#CFLAGS=|g\" aes/Makefile")
-set(BUILD_CMD_STR -j${PARALLEL_JOBS} -C aes CC=${CC} LDFLAGS=${LDFLAGS} AR=${AR} RANLIB=${RANLIB})
+if ($ENV{DARWIN})
+    set(BUILD_CMD_STR -j${PARALLEL_JOBS} -C aes CC=${CC} CFLAGS=${CFLAGS} LDFLAGS=${LDFLAGS} AR=${AR} RANLIB=${RANLIB})
+else()
+    set(BUILD_CMD_STR -j${PARALLEL_JOBS} -C aes CC=${CC} LDFLAGS=${LDFLAGS} AR=${AR} RANLIB=${RANLIB})
+endif()
```
Comment (NiLuJe) [inline thirdparty/minizip/CMakeLists.txt:32]: You can probably drop the DARWIN branch and simply always pass CFLAGS, it's probably what it should already have been doing anyway ;).

Comment (JasonInOttawa) [reply thirdparty/minizip/CMakeLists.txt:32]: I certainly don't mind doing that. Are there any concerns about breaking the build on other platforms? I've only built this for the Mac desktop.

Comment (Frenzie) [reply thirdparty/minizip/CMakeLists.txt:32]: Btw, you should also add it on line 12 or so in that case, and I imagine it's not passed on from Makefile.third either? But agreed it should probably be passed anyway.

Comment (JasonInOttawa) [reply thirdparty/minizip/CMakeLists.txt]: I've removed the DARWIN checks. Is it still necessary to add at line 12 or look at Makefile.third?

Comment (NiLuJe) [reply thirdparty/minizip/CMakeLists.txt]: Looking at it, it appears to go through a dirty, dirty hack: it's passing `CFLAGS` via `CC`. That may be on purpose (i.e., the Makefiles are broken), in which case, you probably don't even need the `BUILD_CMD_STR` bit...

Comment (NiLuJe) [reply thirdparty/minizip/CMakeLists.txt]: Yep, the Makefile doesn't honor CFLAGS.

TL;DR:

* You don't have to tweak our Makefile.third
* You can also stop passing CFLAGS to BUILD_CMD_STR

I have no idea if updating CFLAGS in CMake will magically be reflected in CC, but, oh, well.

Comment (Frenzie) [reply thirdparty/minizip/CMakeLists.txt]: @JasonInOttawa Comments on specific lines are just that! We were just talking about this one specific if DARWIN, not all of them. ;-)

File: thirdparty/minizip/CMakeLists.txt [hunk 2]
```diff
@@ -13,14 +13,17 @@ assert_var_defined(AR)
 assert_var_defined(LDFLAGS)
```
Comment (Frenzie) [inline thirdparty/minizip/CMakeLists.txt:13]: Also, I meant like this. ^_^ (But see @NiLuJe's comment https://github.com/koreader/koreader-base/pull/1198#discussion_r493751534.)
```suggestion
assert_var_defined(CFLAGS)
assert_var_defined(LDFLAGS)
```

**Comment by JasonInOttawa** [comment]
I just did a koreader clone/gh pr checkout 1198 and the build worked. However, running the resulting KOReader.app fails with this error:

ffi.load: libs/libfreetype.6.dylib
failed to run lua chunk: frontend/ui/widget/imagewidget.lua:172: 'struct BlitBuffer8' has no member named 'stride'

I don't believe this has anything to do with this PR. Would it be worthwhile to raise it as an issue?

Thanks!

**Comment by Frenzie** [comment]
That means you've updated front without updating base. I tend to use `git submodule update --recursive`, but if you're working on this PR you may not want that.

What you want for this branch is probably a rebase of your commits on top of current master, e.g., something like `git pull -r upstream master`, but I'd try that first in a throwaway copy branch. ;-)

**Comment by JasonInOttawa** [comment]
Thank you for your reply. I did do this in a throwaway copy, only intending to build koreader including this PR in base:

```
export PATH="/usr/local/opt/gettext/bin:${PATH}"
export PATH="/usr/local/bin:/usr/local/sbin:${PATH/:\/usr\/local\/bin/}"
export MACOSX_DEPLOYMENT_TARGET=10.14
git clone https://github.com/koreader/koreader.git
cd koreader && ./kodev fetch-thirdparty
cd base
gh pr checkout 1198
cd ..
./kodev release macos
... build runs ...
./KOReader-2020.09-30.app/Contents/MacOS/koreader
... log starts then run fails with this error:
ffi.load: libs/libfreetype.6.dylib
failed to run lua chunk: frontend/ui/widget/imagewidget.lua:172: 'struct BlitBuffer8' has no member named 'stride'
```

Is it still possible that I didn't update base, since this is a fresh clone?

Thanks!

**Comment by Galunid** [comment]
Your base is based on old base (before some PR got merged). If you go to https://github.com/JasonInOttawa/koreader-base/tree/add_allow_impl_fn_decl
You will see
image
You can fix it by doing something like this:
```
git checkout master
git pull upstream master -- or however your remote is called
git checkout add_allow_impl_fn_decl
git rebase master
```
Hopefully it goes without any conflicts ;-)

**Comment by JasonInOttawa** [comment]
That worked! Thank you. Only difference is that my remote is called origin

**Comment by JasonInOttawa** [comment]
I hope you don't mind another question. When I look at 

https://github.com/JasonInOttawa/koreader-base/tree/add_allow_impl_fn_decl

I still see that my fork is 7 commits behind master. Is there any way to sync it up and leave my PR in place?

Thanks!

**Comment by ezdiy** [comment]
If you rebased, do `git push`. If that says error (i rebase doesn't stack commits in-order of your past PR), do a `git push --force`. That one will rewrite the PR to rebased tree.

**Comment by JasonInOttawa** [comment]
Thank you!

**Comment by JasonInOttawa** [comment]
It seems that a build is completing as part of commits/pushes. Is it possible to download a copy of the resulting Mac application? Thanks.

**Comment by Frenzie** [comment]
Most likely, please look in the docs for artifact and link it here.

**Comment by NiLuJe** [comment]
I *think* we're missing an artifact upload action for that to behave, but I'm *extremely* unfamiliar with GH Actions.

**Comment by Frenzie** [comment]
Also to be clear this doesn't build an application, just the libraries.

> I think we're missing an upload-artifact action for that to behave, but I'm extremely unfamiliar with GH Actions.

Not entirely sure what you mean, of course we're "missing" it *now*. :-P

**Comment by JasonInOttawa** [comment]
I was actually hoping to find a built KOReader.app Mac application bundle. I may be the only person in the world looking for one, though, so I'll keep downloading and building every release or so :)

I did look at documentation and did a bit of reading about github artifacts, but not enough to actually understand them.

Thanks for all your help!

**Comment by Frenzie** [comment]
That's presumably possible, but not in this repo. But if you've built it once then consecutive builds should take very little time since most things don't have to be recompiled?

**Comment by JasonInOttawa** [comment]
I was looking for artifacts in the koreader repository.

Subsequent builds are much faster, but if I have trouble building I do a fresh clone and reapply the change in this PR. It still only takes about an hour even when fresh.

**Comment by JasonInOttawa** [comment]
Is this change ready for review or merge? I've used it on the last couple releases and it is working for me.

**Comment by NiLuJe** [comment]
At a very quick glance, https://github.com/koreader/koreader-base/pull/1198#discussion_r493751534 is still relevant (minizip's own buildsystem doesn't honor CFLAGS).

**Comment by JasonInOttawa** [comment]
Thank you. I'll look into that.

File: thirdparty/minizip/CMakeLists.txt (hunk 2 above)
Comment (Frenzie) [reply thirdparty/minizip/CMakeLists.txt:14]: @JasonInOttawa There's still a standing review comment! ;-)

File: thirdparty/minizip/CMakeLists.txt (hunk 1 above)
Comment (JasonInOttawa) [reply thirdparty/minizip/CMakeLists.txt]: Thank you. Revising BUILD_CMD_STR from this

`set(BUILD_CMD_STR -j${PARALLEL_JOBS} -C aes CC=${CC} CFLAGS=${CFLAGS} LDFLAGS=${LDFLAGS} AR=${AR} RANLIB=${RANLIB})
`
to this

`set(BUILD_CMD_STR -j${PARALLEL_JOBS} -C aes CC=${CC} LDFLAGS=${LDFLAGS} AR=${AR} RANLIB=${RANLIB})
`

results in the implicit function declaration error. I apologize if I misunderstood the comment, but if I did understand then passing CFLAGS in BUILD_CMD_STR is required for the build to work.

Comment (NiLuJe) [reply thirdparty/minizip/CMakeLists.txt]: You'll want to tweak `CC` (which already includes CFLAGS), c.f., https://github.com/koreader/koreader-base/blob/7231add42900cb5879e1e0e7d1c98aec4f71112c/Makefile.third#L108

Comment (JasonInOttawa) [reply thirdparty/minizip/CMakeLists.txt]: By "tweak CC" do you mean add the flag to the CC variable before setting BUILD_CMD_STR? Or should this change be done in Makefile.third or Makefile.def? Or am I chasing completely the wrong cat?

```
set(CC "${CC} -Wno-error=implicit-function-declaration")
set(BUILD_CMD_STR -j${PARALLEL_JOBS} -C aes CC=${CC} LDFLAGS=${LDFLAGS} AR=${AR} RANLIB=${RANLIB})
```

Thank you!

Comment (NiLuJe) [reply thirdparty/minizip/CMakeLists.txt]: Yep, that should work ;).

Comment (JasonInOttawa) [reply thirdparty/minizip/CMakeLists.txt]: The change is in. Thank you for your help!

File: thirdparty/minizip/CMakeLists.txt (hunk 2 above)
Comment (JasonInOttawa) [reply thirdparty/minizip/CMakeLists.txt]: Thank you!

**Comment by JasonInOttawa** [comment]
I believe the CFLAGS minizip issue is fixed in the latest commit. Is this looking ready for merge? Thanks!

**Comment by NiLuJe** [comment]
Yup, that'll have to do until someone figure out where the `-Werror` came from ;).

**Comment by JasonInOttawa** [comment]
Great. I'll give it a build and see how it goes

**Comment by JasonInOttawa** [comment]
When I do a 

```
git clone https://github.com/koreader/koreader.git
cd koreader && ./kodev fetch-thirdparty
```

I don't see this change. Should I do something differently, or should I just wait a bit?

**Comment by NiLuJe** [comment]
base needs to be bumped in the main repo (or you need to wedge a custom base checkout somewhere in there).

**Comment by JasonInOttawa** [comment]
Thank you. I'll do a checkout for now and keep an eye out for the next base bump.
